### PR TITLE
PROD-10363 : Automate creation of adt instance 

### PR DESCRIPTION
**Pull request by johnfolly**
# BABYLON AZURE ADT

Allow communication with Azure Digital Twin

## Usage:

```bash
babylon azure adt [OPTIONS] COMMAND [ARGS]...
```

### Options:

Name | Type | Description | Default
------------ | ------------- | ------------- | ------------- |
`--help` | **boolean** | Show this message 👆 and exit. | False

**Subcommands**

`instance`: Subgroup dedicate to Azure digital twins instance management
`model`: Subgroup dedicate to Azure digital twins models management

## BABYLON AZURE ADT INSTANCE
Subgroup dedicate to Azure digital twins instance management

Requires key `azure_subscription` in the `platform config` file.

### Usage:

```bash
babylon azure adt instance [OPTIONS] COMMAND [ARGS]...

```
### Options:
Name | Type | Description | Default
------------ | ------------- | ------------- | ------------- |
`--help` | **boolean** | Show this message 👆 and exit. | False

`create`: Create a new ADT instance in current platform resource group
`delete`: Delete a ADT instance in current platform resource group
`get`: Get an azure digital twins instance details
`list`: Get all azure digital twins instances

###  babylon azure adt instance create

Create a new ADT instance in current platform resource group

Requires key `resources_location` in the `platform config` file.

Requires key `resource_group_name` in the `platform config` file.

#### Usage:

```bash
babylon azure adt instance create [OPTIONS] ADT_INSTANCE_NAME
```

#### Options:

Name | Type | Description | Default
------------ | ------------- | ------------- | ------------- |
`--help` | **boolean** | Show this message 👆 and exit. | False

### babylon azure adt instance delete

Delete a ADT instance in current platform resource group

Requires key `resource_group_name` in the `platform config` file.

### Usage:

```bash
babylon azure adt instance delete [OPTIONS] ADT_INSTANCE_NAME
```

#### Options:

Name | Type | Description | Default
------------ | ------------- | ------------- | ------------- |
`-f`, `--force` | **boolean** | Don't ask for validation before delete | False
`--help` | **boolean** | Show this message 👆 and exit. | False

### babylon azure adt instance get
Get an azure digital twins instance details

Requires key `resource_group_name` in the `platform config` file.

#### Usage:

```bash
babylon azure adt instance get [OPTIONS] ADT_INSTANCE_NAME
```

#### Options:

Name | Type | Description | Default
------------ | ------------- | ------------- | ------------- |
`-o`, `--output_file` | **path** | Write full output of the adt api in target file | None
`--help` | **boolean** | Show this message 👆 and exit. | False

### babylon azure adt instance list
Get all azure digital twins instances

Requires key `resource_group_name` in the `platform config` file.

#### Usage:

```bash
babylon azure adt instance list [OPTIONS]
```

#### Options:

Name | Type | Description | Default
------------ | ------------- | ------------- | ------------- |
`-o`, `--output_file` | **path** | Write full output of the adt api in target file | None
`-f`, `--fields` | text | Fields witch will be keep in response data, by default all | None
`--help` | **boolean** | Show this message 👆 and exit. | False

### CODE REVIEW

**Comment by johnfolly** [comment]
# Note

It is not necessary to review `azure adt model` commands i just moved them form `babylon->azure->adt->commands` to  `babylon->azure->adt->groups->model->commands`

File: Babylon/groups/azure/groups/adt/groups/instance/commands/list.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    "-f",
+    "--fields",
+    "fields",
+    type=str,
+    help="Fields witch will be keep in response data, by default all",
+)
+def list(
+    digital_twins_client: AzureDigitalTwinsManagementClient,
+    resource_group_name: str,
+    output_file: Optional[pathlib.Path],
```
Comment (o-reo) [inline Babylon/groups/azure/groups/adt/groups/instance/commands/list.py:46]:
```suggestion
    output_file: Optional[pathlib.Path] = None,
```

File: Babylon/groups/azure/groups/adt/groups/model/commands/upload.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    return True
+
+
+@command()
+@pass_dt_client
+@argument("model_file",
+          type=click.Path(exists=True, file_okay=True, dir_okay=False, readable=True, path_type=pathlib.Path))
+@option("-o", "--override", "override_if_exists", is_flag=True, help="Override existing models")
+@allow_dry_run
+def upload(dt_client: DigitalTwinsClient, model_file: pathlib.Path, override_if_exists: bool, dry_run: bool):
```
Comment (o-reo) [inline Babylon/groups/azure/groups/adt/groups/model/commands/upload.py:65]:
```suggestion
def upload(dt_client: DigitalTwinsClient, model_file: pathlib.Path, override_if_exists: bool = False, dry_run: bool = False):
```

File: Babylon/groups/azure/groups/adt/groups/model/commands/upload.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    """Upload MODEL_FILE to adt
+
+    MODEL_FILE must be a json file"""
+
+    if model_file.suffix != ".json":
+        logger.error(f"{model_file} is not a json file")
+        return
+
+    model_file_content = json.load(open(model_file, "r"))
```
Comment (o-reo) [inline Babylon/groups/azure/groups/adt/groups/model/commands/upload.py:74]:
```suggestion
with open(model_file, "r')) as file:
    model_file_content = json.load(file)
```

**Review by o-reo — COMMENTED** [review]
Thanks !  Just noticed you don't put default value to optional function arguments. Typing does not have any runtime effect in Python so it's not really optional before putting a default value.
The `click` command does allow this but we may have to update the code at one point to not use click.

**Review by o-reo — APPROVED** [review]
All good

**Pull request merged**